docker/witnet-rust/wip0010-recovery.sh

```diff
@@ -0,0 +1,156 @@
+  log "Skipping WIP-0010 recovery"
+  exit 0
+fi
+
+# Make sure the arguments make sense
+if ! command -v "$WITNET_BINARY" &> /dev/null; then
+  log "ERROR: The provided witnet binary (first argument to this script) is not a valid executable file: $WITNET_BINARY"
+  exit 1
+fi
+if [ ! -f "$WITNET_CONFIG_FILE" ]; then
+  log "ERROR: The provided witnet configuration file (second argument to this script) is not a valid configuration file: $WITNET_CONFIG_FILE"
+  exit 2
+fi
+
+# Read configuration (e.g. node server address) from config file
+log "Using configuration file at $WITNET_CONFIG_FILE"
+HOST=$(grep "server_addr" "$WITNET_CONFIG_FILE" | sed -En "s/server_addr = \"(.*)\"/\1/p" | sed -E "s/0\.0\.0.\0/127.0.0.1/g" )
+ADDRESS=$(echo "$HOST" | cut -d':' -f1)
+PORT=$(echo "$HOST" | cut -d':' -f2)
+
+# Check connection to local witnet node
+TIME_TO_NEXT_RETRY=5
+log "Checking connection to local witnet node at $HOST"
+while true
+  if nc -zv "$ADDRESS" "$PORT" &>/dev/null; then
+    log "Successful connection to local witnet node at $HOST"
+    break
+  else
+    log "ERROR: Failed to connect to local witnet node at $HOST"
+    log "Retrying in $TIME_TO_NEXT_RETRY seconds"
+    sleep "$TIME_TO_NEXT_RETRY"
+    TIME_TO_NEXT_RETRY=$(( 2 * TIME_TO_NEXT_RETRY ))
+  fi
+do true; done
+
+# Check whether the local witnet node is below WIP-0010 "common checkpoint" (#248839)
+if $WITNET_BINARY node blockchain --epoch 248839 --limit 1 2>&1 | grep -q "block for epoch"; then
+  log "The local witnet node at $HOST seems to be syncing blocks prior to the WIP-0010 fork. No recovery action is needed"
+  exit 0
+fi
+
+# Check whether the local witnet node is on the `A` chain, and if so, skip recovery
+if $WITNET_BINARY node blockchain --epoch 248839 --limit 2 2>&1 | grep -q '#248921 had digest 7556670d'; then
+  log "The local witnet node at $HOST seems to be on the leading chain. No recovery action is needed"
+  exit 0
+fi
+
+# There is no way back, recovery is needed
+echo "$RECOVERY_BANNER"
+
+# Update known peers in configuration file
+log "Updating known_peers in configuration file at $WITNET_CONFIG_FILE"
+sed -ziE "s/known_peers\s*=\s*\[\n.*\\,\n\]/known_peers = [$KNOWN_PEERS]/g"  "$WITNET_CONFIG_FILE" &&
+log "Successfully updated known_peers in configuration file" ||
+log "ERROR: Failed to update known_peers in configuration file at $WITNET_CONFIG_FILE"
+
+# Rewind local chain back to the WIP-0010 "common checkpoint" (#248839)
+log "Triggering rewind of local block chain back to epoch #248839"
+$WITNET_BINARY node rewind --epoch 248839 &>/dev/null
+REWIND_START=$(date +%s)
+
+# Flush existing peers and inject new peers in runtime
+$WITNET_BINARY node clearPeers 2>&1 | grep -q "Successful" &&
+log "Successfully cleared existing peers from buckets" ||
+log "ERROR: Failed to clear existing peers from buckets"
+printf "%b" "$KNOWN_PEERS" | sed -En "s/\s*\"(.*)\"\,/\1/p" | "$WITNET_BINARY" node addPeers 2>&1 | grep -q "Successful" &&
```

Comment:
This doesn't work because `addPeers` processes the input line by line, so when the first line is "Successful", grep finds a match and kills the process. The result is that only the first 2 or 3 addresses are added.

Solution: if `$KNOWN_PEERS` was a single line, all the addresses from that line would be added at once.
Solution 2: grep should have a flag to keep searching after the first match

Reply:
I made it totally best-effort, but yeah, it's totally worth it putting it in a single line. Is some separator needed? Whitespace? Comma?

Reply:
You can use any separator except `.` and `:`, use your favorite format.

Reply:
Thanks! Done!